lib/Dialect/FIRRTL/Transforms/CheckCombLoops.cpp

```diff
@@ -0,0 +1,408 @@
+        return failure();
+    }
+    visiting.erase(node);
+    // Finished discovering all the reachable nodes from node.
+    visited.insert(node);
+    return success();
+  }
+
+  /// Traverse all the reachable paths from InstanceOp and MemOps.
+  bool handleDefOps(Node &node, Node &inputArg) {
+    if (node.val.isa<BlockArgument>())
+      return false;
+
+    return TypeSwitch<Operation *, bool>(node.val.getDefiningOp())
+        .Case<InstanceOp>([&](InstanceOp ins) {
+          size_t portNum = 0;
+          // Get the port index for `node.val`.
+          for (const auto &p : llvm::enumerate(ins->getOpResults()))
+            if (p.value() == node.val) {
+              portNum = p.index();
+              break;
+            }
+          FModuleOp referencedModule =
+              dyn_cast<FModuleOp>(*instanceGraph.getReferencedModule(ins));
+          if (referencedModule) {
+            auto port = referencedModule.getBodyBlock()->getArgument(portNum);
+            // Get all the ports that have a combination path from `port`.
+            for (const auto &portNode : portPaths[Node(port, node.fieldId)]) {
+              auto instResult = ins.getResult(
+                  portNode.val.cast<BlockArgument>().getArgNumber());
+              Node instNode(instResult, portNode.fieldId);
+              if (dfsFromNode(instNode, inputArg).failed()) {
+                auto remark =
+                    ins.emitRemark("instance is part of a combinational "
+                                   "cycle, instance port number '")
+                    << portNode.val.cast<BlockArgument>().getArgNumber()
+                    << "' has a path to port number '" << portNum << "'"
+                    << " " << ins.getName() << "."
+                    << ins.getPortName(
+                              portNode.val.cast<BlockArgument>().getArgNumber())
+                           .getValue();
+                if (auto nodeType =
+                        portNode.val.getType().dyn_cast<BundleType>()) {
+
+                  remark << "."
+                         << nodeType.getElementName(
+                                nodeType.getIndexForFieldID(portNode.fieldId));
+                }
+                remark << " --> " << ins.getName() << "."
+                       << ins.getPortName(portNum).getValue();
+                if (auto nodeType = node.val.getType().dyn_cast<BundleType>()) {
+                  remark << "."
+                         << nodeType.getElementName(
+                                nodeType.getIndexForFieldID(node.fieldId));
+                }
+                return true;
+              }
+            }
+          }
+          return false;
+        })
+        .Case<RegOp, RegResetOp>([&](auto) { return true; })
+        .Case<MemOp>([&](MemOp mem) {
+          auto type = node.val.getType().cast<BundleType>();
+          auto enableFieldId = type.getFieldID(1);
+          auto dataFieldId = type.getFieldID(3);
```

Comment:
Is there any helper function to avoid hardcoding of index of enable and data field?

Reply:
Thanks for spotting this, I didn't realize enums exist. Fixed it.